airflow/executors/knative_executor.py

```diff
@@ -0,0 +1,186 @@
+        :type command: str
+        """
+        if key is None:
+            return
+        self.log.info("%s running %s", self.__class__.__name__, key)
+        (dag_id, task_id, execution_date, try_number) = key
+        try:
+            future = make_request_async(
+                task_id,
+                dag_id,
+                execution_date,
+                host=self.host, log=self.log, host_header=self.host_header)
+            resp = await future
+            if resp.status != 200:
+                state = State.FAILED
+                self.log.error("Failed to execute task %s.", str(resp.text))
+                self.result_pipe.send((key, state))
+            else:
+                self.log.info("assuming task success")
+                # self.result_queue.put((key, None))
+                self.pop_running_pipe.send(key)
+
+        except asyncio.InvalidStateError as e:
+            state = State.FAILED
+            self.log.error("Failed to execute task %s.", str(e))
+            self.result_pipe.send((key, state))
+
+
+class KnativeExecutor(BaseExecutor):
+    """
+    KnativeExecutor executes tasks locally in parallel. It uses the
+    multiprocessing Python library and queues to parallelize the execution
+    of tasks.
+    """
+    task_queue = None
+    result_queue = None
+    result_pipe = None
+    pop_running_queue = None
+
+    def terminate(self):
+        pass
+
+    @cached_property
+    def kube_executor(self):
+        self.kube_executor_initialized = True
+        return KubernetesExecutor()
+
+    def __init__(self):
+        self.kube_executor_initialized = False
+        super().__init__()
+
+    def start(self):
+        req = conf.get("knative", "knative_host")
+        host_header = None
+        if req is None:
+            raise AirflowException("you must set a knative host")
+
+        self.result_pipe, child_result_pipe = multiprocessing.Pipe()
+        self.task_pipe, child_task_pipe = multiprocessing.Pipe()
+        self.pop_running_pipe, child_pop_running_pipe = multiprocessing.Pipe()
+
+        self.local_loop = KnativeRequestLoop(
+            task_pipe=child_task_pipe,
+            result_pipe=child_result_pipe,
+            host=req,
+            host_header=host_header,
+            pop_running_pipe=child_pop_running_pipe
+        )
+        self.local_loop.start()
+
+    def execute_async(self, key, command, queue=None, executor_config=None):
+        if queue == "kubernetes":
+            self.kube_executor.execute_async(key, command, queue, executor_config)
+        else:
+            self.task_pipe.send(key)
+
+    def check_heartbeats(self):
+        pass
+
+    def sync(self):
+        self.check_heartbeats()
+        while self.result_pipe.poll():
+            results = self.result_pipe.recv()
+            (_, state) = results
+            self.change_state(*results)
+        while self.pop_running_pipe.poll():
+            self.set_not_running(self.pop_running_pipe.recv())
```

Comment:
This won't put anything in the executor event queue, which might confuse the Scheduler. Unless this is _required_ I think just sending a SUCCESS over the result_pipe would be better.

(I _think_ that all the other executors return Success even when the tak might fail -- they just treat it as "did we exit cleanly" and the status of the task doesn't affect that)

Reply:
But wouldn't sending a SUCCESS here create false positives? If the executor recieves a 200 then the task has completed remotely. I believe this is similar to how the k8sexecutor works.

Reply:
It seems the k8sexecutor gives a state of None in the case of successes

Reply:
I'm not quite sure _what_ the executor events are used for in the scheduler - I haven't gotten to the bottom of that yet, but I feel we should probably copy what Kube executor does and send _some_ event.